src-cpp/rdkafkacpp_int.h

```diff
@@ -213,8 +337,18 @@ class MessageImpl : public Message {
    * used as a place holder and rkmessage_ is set to point to it. */
   rd_kafka_message_t rkmessage_err_;
   mutable std::string *key_; /* mutable because it's a cached value */
+  RdKafka::Headers *headers_;
 
 private:
+  RdKafka::Headers* get_headers_from_rkmessage(rd_kafka_message_t *rkmessage) {
+    rd_kafka_headers_t *hdrsp;
+    rd_kafka_resp_err_t err;
+
+    if (rkmessage->len > 0 && !(err = rd_kafka_message_headers(rkmessage, &hdrsp))) {
```

Comment:
you need to detach the headers to make sure they dont get freed with the rkmessage